File: 02_hgf_modeling/bms_vch.py
"""
bms_vch.py  ─  RFX-BMS comparing N VCH HGF model variants
============================================================

Implements Random-Effects Bayesian Model Selection (Stephan et al., 2009,
NeuroImage 46:1004–1017) with Protected Exceedance Probabilities (Rigoux
et al., 2014, NeuroImage 84:971–985) in pure Python/NumPy.

Algorithm overview
------------------
Given per-subject log model evidence LME_nm (subject n, model m):

1. Variational E/M updates on a Dirichlet distribution over model frequencies r:
     E-step:  u_nm  ∝ exp( LME_nm + ψ(α_m) − ψ(Σα) )   [normalised across m]
     M-step:  α_m   =  α0 + Σ_n u_nm
   Iterate to convergence.  Starting α0 = 1 (uniform Dirichlet prior).

2. Expected frequency:  Ef_m = α_m / Σα

3. Exceedance probability XP_m = P(r_m > r_j ∀j≠m | data)
   Estimated by drawing 10⁶ samples from Dirichlet(α).

4. Bayesian Omnibus Risk (BOR, Rigoux et al. 2014):
     log p(y|H0) = Σ_n log[ (1/K) Σ_m exp(LME_nm) ]   [null: all models equal]
     log p(y|H1) ≈ F_VB                                  [VB free energy]
     BOR = sigmoid( log p(H0|y) )   [equal priors on H0, H1]

5. Protected XP:  PXP_m = (1 − BOR)·XP_m + BOR/K

Sections controlled by toggle flags:
  DO_RUN   — launch bms_vch.jl locally (skips participants with existing output)
  DO_BMS   — load LME CSVs and run RFX-BMS
  DO_PLOT  — bar charts of Ef and PXP
"""

################################################################################
# ─── ROLE IN THE MANUSCRIPT ───────────────────────────────────────────────────
#
# MODEL SELECTION, part 2 of 2.  Produces Supplementary Fig. S3c.
#
#     bms_vch.jl (per-subject evidence) ──▶ THIS SCRIPT (group-level RFX-BMS)
#
# Why random-effects BMS rather than a fixed-effects comparison
#     Fixed-effects comparison assumes one model generated everyone's data, so a
#     handful of subjects with extreme evidence can dominate the group result.
#     RFX-BMS instead treats model identity as a random variable across subjects
#     and estimates the population frequency of each model, which is the right
#     assumption for a clinical sample that may well be heterogeneous.
#
#     Protected exceedance probability (PXP) guards the headline number against
#     the possibility that no model is genuinely better than any other: it mixes
#     the raw exceedance probability with the uniform null in proportion to the
#     Bayesian Omnibus Risk.  A low BOR means the models really do differ.
#
# Reproducibility note
#     bms_vch.jl does not set a random seed for its MAP random-restart fallback,
#     so its per-subject evidence output — and therefore the Ef/PXP/BOR this
#     script computes from it — varies slightly between runs. The XP-sampling
#     step below (n_xp_samples draws from Dirichlet(alpha)) is itself seeded
#     (default seed=42), but that only fixes the sampling noise on top of
#     whatever per-subject evidence bms_vch.jl produced that run.
#
# Output
#     model_comparison/bms/bms_summary.csv  and  figures/bms_ef_pxp.png.
#     The published panel was re-rendered from that CSV, with only cosmetic
#     changes, by hppd_manuscript_master/03_visualization/supplement/
#     hgf_bms_modified.py.
################################################################################

import os
import sys
import glob
import subprocess
import numpy as np
import pandas as pd
from scipy.special import digamma, gammaln, logsumexp
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import seaborn as sns

################################################################################
# ─── FLAGS ────────────────────────────────────────────────────────────────────
################################################################################

DO_RUN  = True    # run bms_vch.jl (skips subjects already done)
DO_BMS  = True    # run RFX-BMS on collected LME CSVs
DO_PLOT = True    # generate Ef + PXP bar charts

################################################################################
# ─── ACTIVE CONFIG ────────────────────────────────────────────────────────────
# MODEL_KEYS: which models to include in the BMS comparison.
#   Any subset of model types that have been computed by bms_vch.jl.
#   Supported strings (must have corresponding lme_{key} column in LME CSVs):
#     "2level_empiric"             — 2-level HGF, current empirical conditions
#     "2level_nominal"             — 2-level HGF, raw condition proportions (0/0.25/0.5/0.75)
#     "3level_empiric"             — 3-level HGF, current empirical conditions
#     "3level_nominal"             — 3-level HGF, raw condition proportions
################################################################################

MODEL_KEYS = ["2level_empiric", "3level_empiric", "2level_nominal", "3level_nominal"]

# Display labels (edit to taste)
MODEL_LABELS = {
    "2level_empiric"             : "2-level\n(empirical)",
    "2level_nominal"             : "2-level\n(nominal)",
    "3level_empiric"             : "3-level\n(empirical)",
    "3level_nominal"             : "3-level\n(nominal)",
}

# Bar colours — cycles if more models than colours
_PALETTE = ["#4472C4", "#ED7D31", "#70AD47", "#FFC000",
            "#5B9BD5", "#C55A11", "#375623"]

################################################################################
# ─── PATHS ────────────────────────────────────────────────────────────────────
################################################################################

SCRIPT_DIR  = os.path.dirname(os.path.abspath(__file__))
BMS_DIR     = os.path.join(SCRIPT_DIR, "model_comparison", "bms")
RESULTS_DIR = os.path.join(BMS_DIR, "results")
FIGURES_DIR = os.path.join(SCRIPT_DIR, "model_comparison", "figures")
os.makedirs(RESULTS_DIR, exist_ok=True)
os.makedirs(FIGURES_DIR, exist_ok=True)

################################################################################
# ─── RFX-BMS IMPLEMENTATION ───────────────────────────────────────────────────
################################################################################

def rfx_bms(log_evidence, alpha0=1.0, max_iter=1000, tol=1e-6,
            n_xp_samples=1_000_000, seed=42):
    """
    Random-Effects Bayesian Model Selection (Stephan et al. 2009) with
    Protected Exceedance Probabilities (Rigoux et al. 2014).

    Parameters
    ----------
    log_evidence : ndarray, shape (n_subjects, n_models)
        Laplace log marginal evidence per subject × model.
    alpha0 : float
        Dirichlet prior concentration parameter (1 = uniform).
    max_iter, tol : VB convergence settings.
    n_xp_samples  : Monte-Carlo samples for XP estimation.
    seed           : RNG seed for reproducibility.

    Returns
    -------
    alpha : ndarray (n_models,) — posterior Dirichlet parameters
    Ef    : ndarray (n_models,) — expected model frequencies
    xp    : ndarray (n_models,) — exceedance probabilities
    pxp   : ndarray (n_models,) — protected exceedance probabilities
    bor   : float               — Bayesian Omnibus Risk
    F     : float               — variational free energy
    """
    n, k = log_evidence.shape
    alpha0_vec = np.ones(k) * alpha0
    alpha = alpha0_vec + n / k

    u = np.zeros((n, k))
    for _ in range(max_iter):
        alpha_old = alpha.copy()
        log_u  = log_evidence + (digamma(alpha) - digamma(alpha.sum()))
        log_u -= log_u.max(axis=1, keepdims=True)
        u      = np.exp(log_u)
        u     /= u.sum(axis=1, keepdims=True)
        alpha = alpha0_vec + u.sum(axis=0)
        if np.max(np.abs(alpha - alpha_old)) < tol:
            break

    Ef = alpha / alpha.sum()

    rng     = np.random.default_rng(seed)
    samples = rng.dirichlet(alpha, size=n_xp_samples)
    winners = samples.argmax(axis=1)
    xp      = np.array([(winners == m).mean() for m in range(k)])

    F_data = (u * log_evidence).sum()
    kl = (gammaln(alpha.sum())    - gammaln(alpha0_vec.sum())
          - gammaln(alpha).sum()  + gammaln(alpha0_vec).sum()
          + ((alpha - alpha0_vec) * (digamma(alpha) - digamma(alpha.sum()))).sum())
    F = F_data - kl

    F0      = (logsumexp(log_evidence, axis=1) - np.log(k)).sum()
    log_bor = F0 - np.logaddexp(F0, F)
    bor     = float(np.clip(np.exp(log_bor), 0.0, 1.0))
    pxp     = (1 - bor) * xp + bor / k

    return alpha, Ef, xp, pxp, bor, F

################################################################################
# ─── DO_RUN ───────────────────────────────────────────────────────────────────
################################################################################

if DO_RUN:
    cmd = ["julia", "--project=.", "--threads", "1", "bms_vch.jl"]
    print(f"\nRunning: {' '.join(cmd)}")
    subprocess.run(cmd, cwd=SCRIPT_DIR, check=True)

################################################################################
# ─── DO_BMS ───────────────────────────────────────────────────────────────────
################################################################################

if DO_BMS:
    lme_files = sorted(glob.glob(os.path.join(RESULTS_DIR, "*_lme.csv")))
    if not lme_files:
        print(f"\n[BMS] No LME files found in {RESULTS_DIR}. Run DO_RUN first.")
    else:
        dfs    = [pd.read_csv(f) for f in lme_files]
        lme_df = pd.concat(dfs, ignore_index=True)

        # Validate that all requested model keys have LME columns
        missing_cols = [k for k in MODEL_KEYS
                        if f"lme_{k}" not in lme_df.columns]
        if missing_cols:
            raise ValueError(
                f"LME columns missing for models: {missing_cols}. "
                f"Re-run bms_vch.jl with these models in MODEL_TYPES."
            )

        # Include only participants who converged for ALL requested models
        conv_cols  = [f"converged_{k}" for k in MODEL_KEYS]
        lme_cols   = [f"lme_{k}"       for k in MODEL_KEYS]
        valid      = lme_df[conv_cols].all(axis=1)
        lme_valid  = lme_df[valid].copy()
        n_excluded = (~valid).sum()

        print(f"\n[BMS] {valid.sum()}/{len(lme_df)} participants with valid LME "
              f"for all {len(MODEL_KEYS)} models ({n_excluded} excluded)")
        if n_excluded > 0:
            print(f"  Excluded IDs: {lme_df[~valid]['record_id'].tolist()}")

        log_ev = lme_valid[lme_cols].values
        alpha, Ef, xp, pxp, bor, F = rfx_bms(log_ev)

        print(f"\n{'─'*60}")
        print(f"  RFX-BMS results  (n = {len(lme_valid)} subjects, "
              f"{len(MODEL_KEYS)} models)")
        print(f"  BOR (Bayesian Omnibus Risk) = {bor:.4f}")
        print(f"  {'Model':<22s}  {'α':>7s}  {'Ef':>7s}  {'XP':>7s}  {'PXP':>7s}")
        for i, key in enumerate(MODEL_KEYS):
            print(f"  {key:<22s}  {alpha[i]:>7.2f}  {Ef[i]:>7.3f}  "
                  f"{xp[i]:>7.3f}  {pxp[i]:>7.3f}")
        print(f"{'─'*60}\n")

        n_models = len(MODEL_KEYS)
        summary = pd.DataFrame({
            "model"      : MODEL_KEYS,
            "alpha"      : alpha,
            "Ef"         : Ef,
            "XP"         : xp,
            "PXP"        : pxp,
            "BOR"        : [bor]      * n_models,
            "F_VB"       : [F]        * n_models,
            "n_subjects" : [len(lme_valid)] * n_models,
        })
        summary_path = os.path.join(BMS_DIR, "bms_summary.csv")
        summary.to_csv(summary_path, index=False)
        print(f"Saved: {summary_path}")

        lme_valid.to_csv(os.path.join(BMS_DIR, "lme_per_subject.csv"), index=False)
        print(f"Saved: {os.path.join(BMS_DIR, 'lme_per_subject.csv')}")

################################################################################
# ─── DO_PLOT ──────────────────────────────────────────────────────────────────
################################################################################

if DO_PLOT:
    matplotlib.rcParams.update({
        "font.family":     "Arial",
        "axes.labelsize":  20,
        "xtick.labelsize": 13,
        "ytick.labelsize": 14,
    })

    summary_path = os.path.join(BMS_DIR, "bms_summary.csv")
    if not os.path.exists(summary_path):
        print("\n[PLOT] No BMS summary found. Run DO_BMS first.")
    else:
        summary = pd.read_csv(summary_path)
        n_models = len(summary)
        bor      = summary["BOR"].iloc[0]
        n_sub    = summary["n_subjects"].iloc[0]
        x        = np.arange(n_models)
        colors   = [_PALETTE[i % len(_PALETTE)] for i in range(n_models)]
        labels   = [MODEL_LABELS.get(k, k) for k in summary["model"]]
        bar_w    = max(0.3, min(0.6, 2.4 / n_models))
        fig_w    = max(9, 2.5 * n_models)

        fig, axes = plt.subplots(1, 2, figsize=(fig_w, 5))

        # ── Ef ────────────────────────────────────────────────────────────────
        axes[0].bar(x, summary["Ef"], color=colors, width=bar_w, edgecolor="none")
        axes[0].axhline(1 / n_models, color="gray", lw=1.2,
                        linestyle="--", alpha=0.7)
        axes[0].set_xticks(x)
        axes[0].set_xticklabels(labels)
        axes[0].set_ylabel("Ef")
        axes[0].set_ylim(0, 1)
        for xi, ef in zip(x, summary["Ef"]):
            axes[0].text(xi, ef + 0.02, f"{ef:.3f}", ha="center", va="bottom",
                         fontsize=11, fontweight="bold")
        sns.despine(ax=axes[0])

        # ── PXP ───────────────────────────────────────────────────────────────
        axes[1].bar(x, summary["PXP"], color=colors, width=bar_w, edgecolor="none")
        axes[1].axhline(1 / n_models, color="gray", lw=1.2,
                        linestyle="--", alpha=0.7)
        axes[1].set_xticks(x)
        axes[1].set_xticklabels(labels)
        axes[1].set_ylabel("PXP")
        axes[1].set_ylim(0, 1)
        for xi, pxp_v in zip(x, summary["PXP"]):
            axes[1].text(xi, pxp_v + 0.02, f"{pxp_v:.3f}", ha="center",
                         va="bottom", fontsize=11, fontweight="bold")
        sns.despine(ax=axes[1])

        fig.suptitle(f"RFX-BMS  |  n = {n_sub}  |  BOR = {bor:.3f}",
                     fontsize=14, y=1.02)
        fig.tight_layout()
        out = os.path.join(FIGURES_DIR, "bms_ef_pxp.png")
        fig.savefig(out, dpi=150, bbox_inches="tight")
        plt.close(fig)
        print(f"\n[PLOT] Saved: {out}")

File: 02_hgf_modeling/bms_vch.jl
### bms_vch.jl  ─  Per-subject Laplace LME for RFX-BMS (N-model)
###
### For each participant with a staged stimulus CSV, computes the MAP estimate
### and Laplace-approximate log marginal evidence (LME) for every model in
### MODEL_TYPES.  Output is used by bms_vch.py for RFX-BMS.
###
### Laplace LME approximation (Stephan et al. 2009, NeuroImage):
###   LME ≈ log p(y,θ|M) + (k/2)·log(2π) − ½·log|H|
###   where H = ForwardDiff.hessian(map_result.f, θ_MAP)  [Hessian of neg-log-joint]
###         k = number of model parameters
###
### MAP warm-start strategy (reduces Laplace failures):
###   1. If MCMC medians file exists for this subject × model: initialize MAP
###      from posterior median (bijected to unconstrained space).
###   2. If median-init fails or no file: try up to MAX_RESTARTS random-restart MAPs.
###   3. Warn if all attempts fail — subject excluded from BMS for that model.
###
### Usage (run locally — MAP is fast, no MCMC):
###   julia --project=. --threads 1 bms_vch.jl
###
### Output: model_comparison/bms/results/<id>_lme.csv
###   Columns: record_id, lme_<model>, converged_<model>  for each model in MODEL_TYPES

################################################################################
# ─── ROLE IN THE MANUSCRIPT ───────────────────────────────────────────────────
#
# MODEL SELECTION, part 1 of 2.  Produces Supplementary Fig. S3c.
#
#     THIS SCRIPT (per-subject evidence) ──▶ bms_vch.py (group-level RFX-BMS)
#
# Why this exists
#     Two modelling choices had to be adjudicated rather than assumed: how deep
#     the hierarchy should be (2-level vs 3-level), and which stimulus convention
#     to feed the HGF (empiric vs nominal detection probabilities).  Crossing
#     those gives the four candidate models compared here.
#
#     This script supplies the per-subject evidence term.  It uses MAP + a Laplace
#     approximation rather than MCMC because model evidence is needed for every
#     subject × every candidate model, and the Laplace route is cheap enough to
#     run locally.  A subject whose Hessian is not positive definite for ANY model
#     in the set is dropped from BMS — that is a limitation of the Laplace
#     approximation at that subject's MAP, and says nothing about the validity of
#     their MCMC estimates, which are retained everywhere else.
#
# Reproducibility note
#     No random seed is set for the random-restart MAP fallback (MAX_RESTARTS,
#     below) or for Julia's default RNG generally. Re-running this script can
#     therefore land on different per-subject MAP estimates than a prior run,
#     which propagates into slightly different Ef/PXP/BOR in bms_vch.py's
#     RFX-BMS output. Set a seed before re-running if you need the result to
#     match a specific prior run exactly.
################################################################################

using ActionModels, HierarchicalGaussianFiltering
using CSV, Glob
using Turing
using Bijectors
using DataFrames
using Distributions
using ForwardDiff
using LinearAlgebra

include("helper_functions/create_agent.jl")

################################################################################
# ─── ACTIVE CONFIG ────────────────────────────────────────────────────────────
# Edit MODEL_TYPES to select which models to compare.
#
# Supported model type strings:
#   "2level_empiric"             — 2-level HGF, corrected empirical mapping (0% → 0.0)
#   "2level_nominal"             — 2-level HGF, raw condition proportions {0,0.25,0.5,0.75}
#   "3level_empiric"             — 3-level HGF, corrected empirical mapping
#   "3level_nominal"             — 3-level HGF, raw condition proportions
#
################################################################################

const MODEL_TYPES = ["2level_empiric", "3level_empiric", "2level_nominal", "3level_nominal"]

# Local base directory for MCMC result files (used for MAP warm-starting).
# Relative to 02_hgf_modeling/ (where this script runs).
# Structure: RESULTS_BASE_DIR/{model_type}/{timepoint}/{modality}/{id}_medians.csv
# Set to nothing to disable median warm-starting (random-restart fallback only).
const RESULTS_BASE_DIR = joinpath("results", "vch")

# Maximum random-restart MAP attempts after median-initialized MAP fails.
const MAX_RESTARTS = 3

################################################################################
# ─── PRIORS ───────────────────────────────────────────────────────────────────
################################################################################

const PRIORS_2LEVEL = (;
    action_precision       = truncated(Normal(0.29350739^(-1), 1), lower = 0.001),
    prior_posterior_weight = truncated(Normal(0.72646851, 1),      lower = 0),
    xprob_volatility       = truncated(Normal(-5.1682685, 1),      upper = -0.5),
)
const PRIORS_3LEVEL = (;
    action_precision       = truncated(Normal(0.29350739^(-1), 1), lower = 0.001),
    prior_posterior_weight = truncated(Normal(0.72646851, 1),      lower = 0),
    xprob_volatility       = truncated(Normal(-5.1682685, 1),      upper = -0.5),
    xvol_volatility        = truncated(Normal(-6, 1),              upper = -0.5),
)

# For model types without dedicated MCMC runs, borrow medians from this
# model type when warm-starting the MAP.
const MEDIAN_FALLBACK = Dict(
    "2level_nominal"             => "2level_empiric",   # borrow 2-level MCMC medians (same architecture)
    "3level_nominal"             => "3level_empiric",   # borrow 3-level MCMC medians (same architecture)
)

################################################################################
# ─── PATHS ────────────────────────────────────────────────────────────────────
################################################################################

const PROJECT   = "hppd_manuscript"
const TIMEPOINT = "hppd"
const MODALITY  = "vch"

################################################################################
# ─── HELPERS ──────────────────────────────────────────────────────────────────
################################################################################

"""Return (hgf_string, priors) for a given model_type."""
function model_config(model_type::String)
    if startswith(model_type, "2level")
        hgf = "binary_2level";  priors = PRIORS_2LEVEL
    elseif startswith(model_type, "3level")
        hgf = "binary_3level";  priors = PRIORS_3LEVEL
    else
        error("Unknown model type: $model_type")
    end
    return hgf, priors
end

"""Return the stimulus vector for a given model_type and data DataFrame."""
function get_stimulus(df::DataFrame, model_type::String)
    if endswith(model_type, "_nominal")
        return collect(Float64, df[!, :condition])
    else
        return collect(Float64, df[!, :empirical_condition])
    end
end

"""
    load_medians(record_id, model_type, priors) → NamedTuple or nothing

Load MCMC posterior medians from disk for warm-starting the MAP optimizer.
Returns a NamedTuple with the same keys as `priors`, or nothing if the file
is missing or unreadable.  Uses MEDIAN_FALLBACK for the nominal variants.
"""
function load_medians(record_id::Int, model_type::String, priors)
    RESULTS_BASE_DIR === nothing && return nothing
    mcmc_type = get(MEDIAN_FALLBACK, model_type, model_type)
    path = joinpath(RESULTS_BASE_DIR, mcmc_type, TIMEPOINT, MODALITY,
                    "$(record_id)_medians.csv")
    isfile(path) || return nothing
    try
        df          = CSV.read(path, DataFrame)
        param_names = keys(priors)
        vals        = [Float64(df[1, string(n)]) for n in param_names]
        return NamedTuple{Tuple(param_names)}(vals)
    catch
        return nothing
    end
end

"""
    medians_to_unconstrained(medians_nt, priors) → Vector{Float64}

Bijected-transform constrained posterior medians to the unconstrained space
that Turing's MAP optimizer operates in.  Uses Bijectors.jl to apply the
same transformation Turing uses internally for each truncated-Normal prior.
Values are clamped 1e-8 inside each bound before transformation to avoid
log(0) at exact boundaries.
"""
function medians_to_unconstrained(medians_nt, priors)
    θ_u = Float64[]
    for n in keys(priors)
        dist = priors[n]
        val  = Float64(medians_nt[n])
        lo, hi = minimum(dist), maximum(dist)
        isfinite(lo) && (val = max(val, lo + 1e-8))
        isfinite(hi) && (val = min(val, hi - 1e-8))
        push!(θ_u, bijector(dist)(val))
    end
    return θ_u
end

"""
    _try_lme(modelfit, k; initial_params=nothing) → (lme, converged)

Single MAP attempt.  Returns (lme::Float64, true) if the Hessian is PD,
(NaN, false) otherwise.  `initial_params` is a vector in unconstrained space.
"""
function _try_lme(modelfit, k; initial_params=nothing)
    if initial_params === nothing
        map_result = maximum_a_posteriori(modelfit.model, adtype=AutoForwardDiff())
    else
        map_result = maximum_a_posteriori(modelfit.model, adtype=AutoForwardDiff(),
                                          initial_params=initial_params)
    end
    θ_u = map_result.optim_result.u
    lp  = -map_result.f(θ_u)
    H   = ForwardDiff.hessian(map_result.f, θ_u)
    if !all(isfinite, H) || !isposdef(Symmetric(H))
        return NaN, false
    end
    lme = lp + (k / 2) * log(2π) - 0.5 * log(det(Symmetric(H)))
    return lme, true
end

"""
    compute_lme(priors, hgf_string, stimuli, responses, record_id, model_type)

Returns (lme::Float64, converged::Bool).

Attempt order:
  1. MAP warm-started from MCMC posterior medians (if file exists).
  2. Up to MAX_RESTARTS random-restart MAPs.
  3. Return (NaN, false) and warn if all attempts fail.
"""
function compute_lme(priors, hgf_string, stimuli, responses,
                     record_id::Int, model_type::String)
    am       = create_agent(hgf_string)
    modelfit = create_model(am, priors, stimuli, responses;
                            check_parameter_rejections=true, verbose=false)
    k = length(priors)

    # ── Attempt 1: warm-start from MCMC posterior medians ─────────────────────
    medians = load_medians(record_id, model_type, priors)
    if medians !== nothing
        try
            θ0       = medians_to_unconstrained(medians, priors)
            lme, conv = _try_lme(modelfit, k; initial_params=θ0)
            conv && return lme, true
        catch e
            @warn "record_id=$record_id model=$model_type: median-init MAP error — $e"
        end
    end

    # ── Fallback: random restarts ──────────────────────────────────────────────
    for attempt in 1:MAX_RESTARTS
        try
            lme, conv = _try_lme(modelfit, k)
            if conv
                note = medians === nothing ? "no medians file" : "median-init non-PD"
                @warn "record_id=$record_id model=$model_type: converged on random restart #$attempt ($note)"
                return lme, true
            end
        catch e
            @warn "record_id=$record_id model=$model_type: random restart #$attempt error — $e"
        end
    end

    n_total = 1 + MAX_RESTARTS   # 1 median-init + MAX_RESTARTS random
    @warn "record_id=$record_id model=$model_type: ALL $n_total MAP attempts failed — excluded from BMS"
    return NaN, false
end

################################################################################
# ─── MAIN ─────────────────────────────────────────────────────────────────────
################################################################################

stim_dir = joinpath("data_n_cmnds", PROJECT, TIMEPOINT, "$(MODALITY)_data")
out_dir  = joinpath("model_comparison", "bms", "results")
mkpath(out_dir)

stim_files = sort(glob("*.csv", stim_dir))
isempty(stim_files) && error("No stimulus CSVs found in $stim_dir")

_model_list = join(MODEL_TYPES, ", ")
println("=" ^ 70)
println("VCH BMS — Laplace LME  ($(length(MODEL_TYPES)) models: $(_model_list))")
println("  $(length(stim_files)) participants")
println("  MCMC warm-start: $(RESULTS_BASE_DIR === nothing ? "disabled" : RESULTS_BASE_DIR)")
println("  Random restarts: $MAX_RESTARTS")
println("=" ^ 70)

n_done = 0
n_skip = 0

for stim_file in stim_files
    record_id = parse(Int, splitext(basename(stim_file))[1])
    out_file  = joinpath(out_dir, "$(record_id)_lme.csv")

    # Determine which models still need computing.
    # If the output file exists but is missing columns for some models (e.g.
    # because MODEL_TYPES was expanded after the original run), read the existing
    # values and compute only the missing models — avoids redundant MAP calls.
    existing_row  = Dict{String, Any}()
    models_to_run = MODEL_TYPES

    if isfile(out_file)
        existing      = CSV.read(out_file, DataFrame)
        models_to_run = [mt for mt in MODEL_TYPES if "lme_$(mt)" ∉ names(existing)]
        if isempty(models_to_run)
            global n_skip += 1
            continue
        end
        # Carry over all existing column values so the merged row is complete
        for col in names(existing)
            existing_row[col] = existing[1, col]
        end
        _models_str = join(models_to_run, ", ")
        println("  →  $record_id  (partial: computing $(_models_str))")
    end

    df        = CSV.read(stim_file, DataFrame)
    responses = collect(Float64, df[!, :response])

    lme_vals  = Dict{String, Float64}()
    conv_vals = Dict{String, Bool}()

    for mt in models_to_run
        hgf_str, priors = model_config(mt)
        stimuli         = get_stimulus(df, mt)
        lme_v, conv_v   = NaN, false
        try
            lme_v, conv_v = compute_lme(priors, hgf_str, stimuli, responses,
                                        record_id, mt)
        catch e
            @warn "record_id=$record_id  model=$mt  unexpected error: $e"
        end
        lme_vals[mt]  = lme_v
        conv_vals[mt] = conv_v
    end

    # Build output row: start from existing values, overwrite/add new model columns
    row = Dict{String, Any}(existing_row)
    row["record_id"] = record_id
    for mt in models_to_run
        row["lme_$(mt)"]       = lme_vals[mt]
        row["converged_$(mt)"] = conv_vals[mt]
    end
    CSV.write(out_file, DataFrame([row]))

    status   = all(conv_vals[mt] for mt in models_to_run) ? "✓" : "!"
    lme_strs = join(["$(mt)=$(isnan(lme_vals[mt]) ? "NaN" : round(lme_vals[mt], digits=1))"
                     for mt in models_to_run], "  ")
    println("  $status  $record_id  $lme_strs")
    global n_done += 1
end

println("\n" * "=" ^ 70)
println("Done: $n_done computed, $n_skip skipped")
println("Results in: $out_dir")
